# Decision Tree Exercises

#### Using Titanic Data
- Remember: your baseline prediction for a classification problem is predicting the most prevelant class in the training dataset (the mode). When you make those predictions, what is your accuracy? This is your baseline accuracy.

In [41]:
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import graphviz
from graphviz import Graph
import prepare

In [23]:
# Get Titanic Data
df = prepare.prep_titanic()
df.head()

passenger_id  survived  pclass     sex  sibsp  parch  embark_town  alone  \
0             0         0       3    male      1      0  Southampton      0   
1             1         1       1  female      1      0    Cherbourg      0   
2             2         1       3  female      0      0  Southampton      1   
3             3         1       1  female      1      0  Southampton      0   
4             4         0       3    male      0      0  Southampton      1   

   sex_male  embark_town_Queenstown  embark_town_Southampton  
0         1                       0                        1  
1         0                       0                        0  
2         0                       0                        1  
3         0                       0                        1  
4         1                       0                        1

In [36]:
#Split Data
train, validate, test = prepare.split_titanic(df)
train.shape, validate.shape, test.shape

((498, 11), (214, 11), (179, 11))

## <font color = 'red'> 1a) What is your baseline prediction?

In [37]:
# Find most often used case in survived
train.survived.value_counts() #most frequent is not_survived

0    307
1    191
Name: survived, dtype: int64

In [ ]:
### Baseline Prediction to 0 or Not_Survived

## <font color = 'red'> 1a) What is your baseline accuracy?

In [38]:
# create copy of df and add baseline column set to 0 or 'not_survived'
train['baseline'] = 0
train.head()

passenger_id  survived  pclass     sex  sibsp  parch  embark_town  alone  \
583           583         0       1    male      0      0    Cherbourg      1   
165           165         1       3    male      0      2  Southampton      0   
50             50         0       3    male      4      1  Southampton      0   
259           259         1       2  female      0      1  Southampton      0   
306           306         1       1  female      0      0    Cherbourg      1   

     sex_male  embark_town_Queenstown  embark_town_Southampton  baseline  
583         1                       0                        0         0  
165         1                       0                        1         0  
50          1                       0                        1         0  
259         0                       0                        1         0  
306         0                       0                        0         0

In [39]:
(train.survived == train.baseline).mean()

0.6164658634538153

### Baseline accuracy is 61.6%

## <font color = 'red'>2) Fit the decision tree classifier to your training sample and transform (i.e. make predictions on the training sample)

In [44]:
# Create X & y version of train, where y is a series with just the target variable and X are all the features. 

X_train = train.drop(columns='survived')
y_train = train.survived

X_validate = validate.drop(columns='survived')
y_validate = validate.survived

X_test = test.drop(columns='survived')
y_test = test.survived

In [45]:
# Create the Decision Tree Object
clf = DecisionTreeClassifier(max_depth=3, random_state=123)

In [46]:
# Fit the Model
clf = clf.fit(X_train, y_train)

ValueError: could not convert string to float: 'male'

In [42]:
dot_data = export_graphviz(clf, feature_names= X_train.columns, class_names=clf.classes_, rounded=True, filled=True, out_file=None)
graph = graphviz.Source(dot_data) 

graph.render('titanic_decision_tree', view=True)

NameError: name 'X_train' is not defined

## <font color = 'red' >3) Evaluate your in-sample results using the model score, confusion matrix, and classification report.

## <font color = 'red'>4) Compute: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support

## <font color = 'red'>5) Run through steps 2-4 using a different max_depth value.

## <font color='red'>6) Which model performs better on your in-sample data?

## <font color = 'red'>7) Which model performs best on your out-of-sample data, the validate set?

## <font color = 'red'>8) Work through these same exercises using the Telco dataset.

## <font color = 'red'>9) Experiment with this model on other datasets with a higher number of output classes.